electrumpersonalserver/server/peertopeer.py

```diff
@@ -0,0 +1,410 @@
+            addr_recv_port = read_int(ptr, payload, 2, False)
+            addr_trans_services = read_int(ptr, payload, 8)
+            addr_trans_ip = payload[ptr[0] : ptr[0]+16]
+            ptr[0] += 16
+            addr_trans_port = read_int(ptr, payload, 2, False)
+            ptr[0] += 8 # skip over nonce
+            user_agent = read_var_str(ptr, payload)
+            start_height = read_int(ptr, payload, 4)
+            if version > RELAY_TX_VERSION:
+                relay = read_int(ptr, payload, 1) != 0
+            else: # must check this node accepts unconfirmed transactions for the broadcast
+                relay = True
+            log.debug(('peer version message: version=%d services=0x%x'
+                + ' timestamp=%s user_agent=%s start_height=%d relay=%i'
+                + ' them=%s:%d us=%s:%d') % (version,
+                services, str(datetime.fromtimestamp(timestamp)),
+                user_agent, start_height, relay, ip_hex_to_str(addr_trans_ip)
+                , addr_trans_port, ip_hex_to_str(addr_recv_ip), addr_recv_port))
+            p2p.sock.sendall(p2p.create_message('verack', b''))
+            self.on_recv_version(p2p, version, services, timestamp,
+                addr_recv_services, addr_recv_ip, addr_trans_services,
+                addr_trans_ip, addr_trans_port, user_agent, start_height,
+                relay)
+        elif command == b'verack':
+            self.on_connected(p2p)
+        elif command == b'ping':
+            p2p.sock.sendall(p2p.create_message('pong', payload))
+
+    # optional override these in a subclass
+
+    def on_recv_version(self, p2p, version, services, timestamp,
+            addr_recv_services, addr_recv_ip, addr_trans_services,
+            addr_trans_ip, addr_trans_port, user_agent, start_height, relay):
+        pass
+
+    def on_connected(self, p2p):
+        pass
+
+    def on_heartbeat(self, p2p):
+        pass
+
+
+class P2PProtocol(object):
+    def __init__(self, p2p_message_handler, remote_hostport=None,
+                 testnet=False, user_agent=DEFAULT_USER_AGENT, relay_txes=False,
+                 socks5_hostport=None, connect_timeout=30, heartbeat_interval=15):
+        '''
+        if remote_hostport = None, use dns_seeds for auto finding peers
+        if socks5_hostport != None, use that proxy 
+        relax_txes controls whether the peer will send you unconfirmed txes
+        heartbeat_interval, how many seconds between heartbeats
+        '''
+        self.log = (log if log else
+                    log.getLogger('ELECTRUMPERSONALSERVER'))
+        self.p2p_message_handler = p2p_message_handler
+        self.testnet = testnet
+        self.user_agent = user_agent
+        self.relay_txes = relay_txes
+        self.socks5_hostport = socks5_hostport
+        self.heartbeat_interval = heartbeat_interval
+        self.connect_timeout = connect_timeout
+        if not self.testnet:  # This seems like a phishing attempt but with the help of a mod?
```

Comment:
Whats the meaning of this comment? Something left in your clipboard perhaps :)

Reply:
Lol weird not sure, I'll remove it.